File: frontend/src/api/rebalancing.js
import api from './index';
import { convertToKRW } from '@/utils/currency.js';

const normalizeAssetType = (value) => {
  if (['주식', '예적금', '현금'].includes(value)) {
    return value;
  }

  return '주식';
};

const normalizeActionColor = (actionType) => {
  if (actionType === 'SELL') {
    return 'border-amber-200 dark:border-amber-500/30 bg-amber-50 dark:bg-amber-500/5 text-amber-600 dark:text-amber-400';
  }

  if (actionType === 'HOLD') {
    return 'border-emerald-200 dark:border-emerald-500/30 bg-emerald-50 dark:bg-emerald-500/5 text-emerald-600 dark:text-emerald-400';
  }

  return 'border-blue-200 dark:border-blue-500/30 bg-blue-50 dark:bg-blue-500/5 text-blue-600 dark:text-blue-400';
};

const buildAssetPayload = (currentAssets) => {
  return currentAssets.map((asset) => ({
    id: asset.id,
    name: asset.name || asset.company_name || asset.ticker || '이름 없음',
    ticker: asset.ticker || asset.name || '',
    type: normalizeAssetType(asset.type),
    quantity: Number(asset.quantity || 0),
    price: Number(asset.price || asset.average_price || 0),
    total: convertToKRW(asset.total, asset.currency),
    currency: 'KRW',
  }));
};

const normalizeRebalancingResponse = (data) => {
  const gaps = Array.isArray(data.gaps) ? data.gaps : [];
  const actions = Array.isArray(data.actions) ? data.actions : [];

  return {
    summary: data.summary || null,
    gapData: gaps.map((item) => ({
      name: item.name,
      현재비중: Number(item.current_weight || 0),
      추천비중: Number(item.target_weight || 0),
      현재금액: Number(item.current_amount || 0),
      목표금액: Number(item.target_amount || 0),
      차이금액: Number(item.difference_amount || 0),
      차이비중: Number(item.gap_weight || 0),
    })),
    actions: actions.map((item) => ({
      id: item.id,
      assetType: item.asset_type,
      actionType: item.action_type,
      title: item.title,
      description: item.description,
      amount: Number(item.amount || 0),
      gapWeight: Number(item.gap_weight || 0),
      priority: item.priority || 'LOW',
      color: normalizeActionColor(item.action_type),
    })),
    logicList: Array.isArray(data.logic_list) ? data.logic_list : [],
    targetWeights: Array.isArray(data.target_weights) ? data.target_weights : [],
  };
};

export const getRebalancingRecommendationAPI = async (
  currentAssets = [],
  marketRegime = 'risk_off'
) => {
  const payload = {
    assets: buildAssetPayload(currentAssets),
    market_regime: marketRegime,
    min_action_amount: 10000,
  };

  const response = await api.post('/api/v1/rebalancing/recommendation', payload);

  return normalizeRebalancingResponse(response.data);
};

File: backend/app/schemas/rebalancing.py
from typing import Literal

from pydantic import BaseModel, Field, field_validator


AssetType = Literal["주식", "예적금", "현금"]
MarketRegime = Literal["risk_off", "neutral", "risk_on"]
ActionType = Literal["BUY", "SELL", "HOLD"]
Priority = Literal["HIGH", "MEDIUM", "LOW"]


class RebalancingAssetInput(BaseModel):
    id: int | str | None = None
    name: str | None = None
    ticker: str | None = None
    type: AssetType = "주식"
    quantity: float = Field(default=0, ge=0)
    price: float = Field(default=0, ge=0)
    total: float | None = Field(default=None, ge=0)
    currency: str = "KRW"

    @field_validator("total", mode="before")
    @classmethod
    def parse_total(cls, value):
        if value is None or value == "":
            return None
        return value


class RebalancingRequest(BaseModel):
    assets: list[RebalancingAssetInput] = Field(default_factory=list)
    market_regime: MarketRegime = "risk_off"
    min_action_amount: float = Field(default=10000, ge=0)


class TargetWeightItem(BaseModel):
    name: AssetType
    weight: float


class RebalancingGapItem(BaseModel):
    name: AssetType
    current_amount: float
    target_amount: float
    difference_amount: float
    current_weight: float
    target_weight: float
    gap_weight: float


class RebalancingActionItem(BaseModel):
    id: int
    asset_type: AssetType
    action_type: ActionType
    title: str
    description: str
    amount: float
    gap_weight: float
    priority: Priority


class RebalancingSummary(BaseModel):
    total_amount: float
    market_regime: MarketRegime
    market_regime_label: str
    risk_score: int
    risk_level: str
    confidence: float
    is_rebalancing_needed: bool


class RebalancingResponse(BaseModel):
    summary: RebalancingSummary
    target_weights: list[TargetWeightItem]
    gaps: list[RebalancingGapItem]
    actions: list[RebalancingActionItem]
    logic_list: list[str]